Update InfluxDB to handle datetime objects and multiple decimal points

philhawthorne opened the pull request:
## Description:
Updates the InfluxDB regex to ignore datetime objects being coverted
into float values.

Adds tests to the component to ensure datetime objects are corectly
handled.

**Related issue (if applicable):** fixes #8042

## Checklist:

If the code does not interact with devices:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

philhawthorne committed:
Update InfluxDB to handle datetime objects

Updates the InfluxDB regex to ignore datetime objects being coverted
into float values.

Adds tests to the component to ensure datetime objects are corectly
handled.

philhawthorne committed:
Fix Hound errors

Fixes errors from Hound bot

philhawthorne committed:
Update InfluxDB to handle multiple decimal points

Changes the way InfluxDB handles values such as 1.2.3.4 to be 1.234 so
it stores in InfluxDB as a valid float value

philhawthorne committed:
Fix lint issues

Reduce the size of a line for the linter

azogue commented:
I can say this is working with the issue I've opened before (#8086), but I have a doubt about:

What need is there for conversion if the string is already being stored?
In many cases, this 'forced' conversion will be generating nonsense data, right?

Would not it be better, simply, to give up conversion in such cases?
```python
except (ValueError, TypeError):
    pass
```

philhawthorne commented:
Hey @azogue 

So the original intention for this was to catch any components that are sending numeric data, but as strings. For example, "45%" instead of "45". When this happens, you can't use the string value in Grafana or other graph suites, as the data stored in InfluxDB isn't a numerical value. 

This change detects values such as "45%", strips out the "%" character and leaves two values, "45" and a _str of "45%". This way, both values are stored in InfluxDB and can be used.

The alternative is each time a new platform comes along, if the developer doesn't return a numerical value for say the battery level, we'll then have to go back and update that component to return the correct data.

azogue commented:
hi @philhawthorne,

I understand the initial intention, and it seems phenomenal to me, but, for the cases of more complex strings with numbers, perhaps the best alternative is to avoid the parsing, right?

For cases of 'number + unit' I think the change is great.

azogue commented on homeassistant/components/influxdb.py:
```diff
@@ -165,9 +171,15 @@ def influx_event_listener(event):
+                            json_body[0]['fields'][key] = float(
+                                non_decimal.sub('', value))
+                        except (ValueError, TypeError):
+                            json_body[0]['fields'][key] = float(
```
@philhawthorne, sorry if I was not understood, I was referring to this except clause, to simply `pass` if a `ValueError` is raised.

philhawthorne committed:
Update InfluxDB to pass on unknown variable

If we get an error trying to convert a variable to a float, let's ignore
it completely

philhawthorne commented:
Thanks @azogue 

That makes more sense. Makes it more simple, and handles both traces that were reported in the issue. I've updated that as you suggested.

Cheers!

azogue reviewed (approved):
Fixes #8086. Looks good to me, and I think it should be included in 0.47.1

mezz64 commented:
Just wanted to confirm that this change fixes the original issue I reported in #8042.

pvizeli reviewed (dismissed):
https://github.com/philhawthorne/home-assistant/blob/74ea240dc05d2d3e1fe6161146f7d0e168946a19/homeassistant/components/influxdb.py#L152-L153 should be as constante on top of component to save compile time every call.

philhawthorne committed:
Make InfluxDB Regex constants

Makes the Regex's used by InfluxDB constants so they don't need to be
compiled each time

philhawthorne commented:
Thanks @pvizeli changed that

pvizeli commented:
I think instead to create this complex things we should use match regex: `^\w*\d+\.?\d+\w*$`

pvizeli committed:
cleanup

pvizeli committed:
fix lint

pvizeli committed:
Update regex

pvizeli commented:
I update the regex for better detection

pvizeli committed:
fix tests

philhawthorne committed:
Fix JSON body missing new line character

philhawthorne commented:
Thanks @pvizeli !

pvizeli committed:
fix exceptions

pvizeli commented:
Good teamwork!

balloob commented:
Cherry-picked for 0.47.1
